sdk/core/core/src/az_span.c

```diff
@@ -176,55 +175,43 @@ AZ_NODISCARD az_result az_span_find(az_span source, az_span target, az_span* out
       {
         // All bytes in `target` have been checked and matched the corresponding bytes in `source` (from the start point `i`),
         // so this is indeed an instance of `target` in that position of `source` (step 4.).
-        // Here we create a span in `source` to map the occurrence of `target` in `source` (pardon the redundancy of words, done for explicit clarity).
-        *out_span = az_span_init(source_ptr + i, target_length, target_length);
 
-        return AZ_OK;
+        return i;
       }
     }
   }
 
   // If the function hasn't returned before, all positions 
   // of `source` have been evaluated but `target` could not be found.
-  return AZ_ERROR_ITEM_NOT_FOUND;
+  const int32_t TARGET_NOT_FOUND = -1;
+
+  return TARGET_NOT_FOUND;
 }
 
 AZ_NODISCARD az_span az_span_token(az_span source, az_span delimiter, az_span* out_remainder)
```

Comment:
If we don't have it in the core SDK, we will need to move it to the iot SDK. 
We need that functionality in IoT SDK.

Reply:
Sounds good. I will look at current IoT usages of this to understand the scenario :)

Reply:
So, I experimented with a couple of things:
https://godbolt.org/z/sn6sZY

If you re-wrote the `az_iot_hub_client_c2d_received_topic_parse` method in terms of the new `az_span_find`, the method size will shrink (assuming all span methods are inlined for simplicity).
- gcc (-O1): 127 to 114 instructions
- clang (-O3): 108 to 84 instructions

Granted, it means a bit more work for you (the caller of `az_span_find` instead of `az_span_token`). I will leave it up to you to decide if its worth the benefit.

https://github.com/Azure/azure-sdk-for-c/blob/35629203f7dab143a65e55760f10ef211f204e1d/sdk/iot/hub/src/az_iot_hub_client_c2d.c#L39-L54

```C
AZ_NODISCARD az_result az_iot_hub_client_c2d_received_topic_parse(
    az_iot_hub_client const* client,
    az_span received_topic,
    az_iot_hub_client_c2d_request* out_request)
{
  AZ_PRECONDITION_NOT_NULL(client);
  AZ_PRECONDITION_VALID_SPAN(received_topic, 1, false);
  AZ_PRECONDITION_NOT_NULL(out_request);

  az_span token;
  token = az_span_slice(received_topic, az_span_find(received_topic, c2d_topic_suffix) + az_span_length(c2d_topic_suffix), -1);
  token = az_span_slice(token, 0, az_span_find(token, c2d_topic_suffix));
  AZ_RETURN_IF_FAILED(az_iot_hub_client_properties_init(&out_request->properties, token));

  return AZ_OK;
}
```

One thing I am not sure about is validation/error checking in the topic parsing between the repeated calls to `az_span_token`. It doesn't look like we ever check if the target string was found or not and assume it will be. Is that expected or should we check the result between the calls to avoid passing in potentially `null` source or -1 for the index?

I was also thinking of mirroring the signature of `az_span_find` and have `az_span_token` return an `int32_t` as well (rather than `az-span`), but that didn't really give great results (not sure why yet, I would have thought doing less work would be good):
```C
AZ_NODISCARD int32_t az_span_token(az_span source, az_span delimiter, az_span* out_remainder);
```

Reply:
Yes, you are absolutely correct, there is validation we need to add in that function.
We didn''t add yet because we are still defining how az_span_find and az_span_token will look like. But it will get added.
So can the function be implemented with az_span_find only, yes, of course it can.
We will verify about code size afterwards, thanks for running that test. I'm not sure we will be able to call the functions inline like that, since we will need to add that validation I spoke above.

Reply:
Would you like me to move az_span_token out of the core SDK for now? Then if its use is proven in the IoT SDK, we can try bringing it back to the core SDK.

Reply:
Yes